### PR TITLE
feat: upsert with on-conflict support

### LINKED ISSUE #139: Upsert on conflict
It should be possible to use upsert `on conflict` in supabase-py/postgrest-py client:

https://supabase.com/docs/reference/javascript/upsert#upsert-your-data

```
const { data, error } = await supabase
  .from('users')
  .upsert({ username: 'supabot' }, { onConflict: 'username' })
```

Right now, if you are upserting but the ID is auto-generated, then you end up with duplicated entries in the table. 
Upsert on conflict allows you to prevent duplicated rows based on other columns.

### PR DESCRIPTION
**Pull request by Oisangui**
- [x] ```on_conflict``` support as described here https://postgrest.org/en/stable/api.html#on-conflict.

Closes #139

### CODE REVIEW

**Review by mansueli — APPROVED** [review]

**Review by lopezjurip — DISMISSED** [review]

File: postgrest/base_request_builder.py
Comment (anand2312) [inline postgrest/base_request_builder.py:72]:
```suggestion
    on_conflict: str= "",
```

File: postgrest/_sync/request_builder.py
Comment (anand2312) [inline postgrest/_sync/request_builder.py:159]: This new arg has to be documented in this method's docstrings so that it shows up on our autodocs

**Review by anand2312 — CHANGES_REQUESTED** [review]

File: postgrest/_sync/request_builder.py
Comment (Oisangui) [reply postgrest/_sync/request_builder.py:159]: Done

**Comment by Oisangui** [comment]
@anand2312 Could you give me a hand telling me what is missing? I'm not very knowledgeable with python repos.

**Comment by troels** [comment]
What is holding this one back? It would be a very nice feature to have!

**Comment by Oisangui** [comment]
> What is holding this one back? It would be a very nice feature to have!
@troels 
Waiting for someone to review the changes that @anand2312 requested.

**Comment by anand2312** [comment]
Sorry for the delay -- can you add some test cases for this case as well please? Otherwise this LGTM, ready to merge @Oisangui

**Comment by J0** [comment]
Think this should be good to go - if merged I'll add test in a FLUP

**Review by J0 — APPROVED** [review]

**Comment by J0** [comment]
Let's also pull in master and make sure the tests pass first though.

**Comment by altryne** [comment]
Would love this to get merged, literally facing this RN and got to this issue. 
The only workaround is SQL func and RPC call? 

CleanShot 2023-02-05 at 21 28 28@2x

**Comment by J0** [comment]
Hey,

@altryne noted - we'll merge by tomorrow evening. Will validate this today evening  + add necessary tests/ fixes.  From now till then then I think the main workaround would be to use a sql func and an RPC call.

**Comment by LeSchuch** [comment]
Hey,

I'm a little bit lost on where to use the on_conflict parameter in a request.
For example, this is not working :
supabase.table("user").upsert([{"id": 5, "name": "John"}], on_conflict=["id"]).execute()

Could you provide a simple example on how to use this parameter ?

**Comment by Oisangui** [comment]
> Hey,
> 
> I'm a little bit lost on where to use the on_conflict parameter in a request. For example, this is not working : supabase.table("user").upsert([{"id": 5, "name": "John"}], on_conflict=["id"]).execute()
> 
> Could you provide a simple example on how to use this parameter ?

You are passing a list, it should be a string.

**Comment by LeSchuch** [comment]
Hum, I have this error even if I'm passing it as a string :
TypeError: SyncRequestBuilder.upsert() got an unexpected keyword argument 'on_conflict'

**Pull request merged**
